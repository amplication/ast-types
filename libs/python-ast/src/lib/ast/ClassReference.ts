import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";

/**
 * Configuration arguments for creating a Python class reference.
 */
export interface ClassReferenceArgs {
  /** The name of the class */
  name: string;
  /** The module name containing the class */
  moduleName?: string;
  /** Optional alias for the class when imported */
  alias?: string;
}

/**
 * Represents a reference to a Python class.
 * This class is used to track and generate imports for classes used in the code.
 *
 * @extends {AstNode}
 */
export class ClassReference extends AstNode {
  /** The name of the class */
  public readonly name: string;
  /** The module name containing the class */
  public readonly moduleName?: string;
  /** Optional alias for the class */
  public readonly alias?: string;

  /**
   * Creates a new Python class reference.
   * @param {ClassReferenceArgs} args - The configuration arguments
   */
  constructor({ name, moduleName, alias }: ClassReferenceArgs) {
    super();
    this.name = name;
    this.moduleName = moduleName;
    this.alias = alias;
  }

  /**
   * Writes the class reference to the writer and adds it to the import collection.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    if (this.moduleName) {
      writer.addImport(this);
    }
    writer.write(this.alias || this.name);
  }
}
