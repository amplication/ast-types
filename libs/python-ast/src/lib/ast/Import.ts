import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";

/**
 * Configuration arguments for creating a Python import statement.
 */
export interface ImportArgs {
  /** The module name to import */
  moduleName: string;
  /** The names to import from the module (if using 'from ... import ...') */
  names?: string[];
  /** The alias for the module (if using 'import ... as ...') */
  alias?: string;
}

/**
 * Represents a Python import statement.
 * This class handles the generation of Python import statements, including
 * simple imports, from imports, and imports with aliases.
 *
 * @extends {AstNode}
 */
export class Import extends AstNode {
  /** The module name to import */
  public readonly moduleName: string;
  /** The names to import from the module */
  public readonly names: string[];
  /** The alias for the module */
  public readonly alias?: string;

  /**
   * Creates a new Python import statement.
   * @param {ImportArgs} args - The configuration arguments
   */
  constructor({ moduleName, names = [], alias }: ImportArgs) {
    super();
    this.moduleName = moduleName;
    this.names = names;
    this.alias = alias;
  }

  /**
   * Writes the import statement to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    writer.addImport(this);
  }
}
