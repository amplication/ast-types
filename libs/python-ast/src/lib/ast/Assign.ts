import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { ClassReference } from "./ClassReference";

/**
 * Configuration arguments for creating a Python assignment statement.
 */
export interface AssignArgs {
  /** The target variable name */
  target: string;
  /** The value to assign */
  value: string;
  /** Type annotation for the target */
  type?: ClassReference;
  /** Whether this is a class variable (vs. instance variable) */
  isClassVariable?: boolean;
}

/**
 * Represents a Python assignment statement.
 * This class handles the generation of Python assignment statements, which can
 * include type annotations and handle class vs. instance variables.
 *
 * @extends {AstNode}
 */
export class Assign extends AstNode {
  /** The target variable name */
  public readonly target: string;
  /** The value to assign */
  public readonly value: string;
  /** Type annotation for the target */
  public readonly type?: ClassReference;
  /** Whether this is a class variable */
  public readonly isClassVariable: boolean;

  /**
   * Creates a new Python assignment statement.
   * @param {AssignArgs} args - The configuration arguments
   */
  constructor({ target, value, type, isClassVariable = false }: AssignArgs) {
    super();
    this.target = target;
    this.value = value;
    this.type = type;
    this.isClassVariable = isClassVariable;
  }

  /**
   * Writes the assignment statement to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    writer.write(this.target);

    if (this.type) {
      writer.write(": ");
      this.type.write(writer);
      writer.write(" = ");
    } else {
      writer.write(" = ");
    }

    writer.write(this.value);
  }
}
