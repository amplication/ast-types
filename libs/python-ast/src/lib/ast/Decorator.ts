import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { ClassReference } from "./ClassReference";

/**
 * Configuration arguments for creating a Python decorator.
 */
export interface DecoratorArgs {
  /** The name of the decorator */
  name: string;
  /** The arguments to pass to the decorator */
  args?: string[];
  /** The keyword arguments to pass to the decorator */
  kwargs?: Record<string, string>;
  /** Optional module name for import */
  moduleName?: string;
}

/**
 * Represents a Python decorator.
 * This class handles the generation of Python decorators, which are used to modify
 * the behavior of functions, methods, and classes.
 *
 * @extends {AstNode}
 */
export class Decorator extends AstNode {
  /** The name of the decorator */
  public readonly name: string;
  /** The arguments to pass to the decorator */
  public readonly args: string[];
  /** The keyword arguments to pass to the decorator */
  public readonly kwargs: Record<string, string>;
  /** The module name containing the decorator */
  public readonly moduleName?: string;
  /** The class reference for import handling */
  private readonly reference?: ClassReference;

  /**
   * Creates a new Python decorator.
   * @param {DecoratorArgs} args - The configuration arguments
   */
  constructor({ name, args = [], kwargs = {}, moduleName }: DecoratorArgs) {
    super();
    this.name = name;
    this.args = args;
    this.kwargs = kwargs;
    this.moduleName = moduleName;

    if (moduleName) {
      this.reference = new ClassReference({
        name: this.name,
        moduleName: this.moduleName,
      });
    }
  }

  /**
   * Writes the decorator to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    writer.write("@");

    if (this.reference) {
      this.reference.write(writer);
    } else {
      writer.write(this.name);
    }

    // Add arguments and keyword arguments if provided
    const hasArgs = this.args.length > 0;
    const hasKwargs = Object.keys(this.kwargs).length > 0;

    if (hasArgs || hasKwargs) {
      writer.write("(");

      // Write positional arguments
      this.args.forEach((arg, index) => {
        writer.write(arg);
        if (index < this.args.length - 1 || hasKwargs) {
          writer.write(", ");
        }
      });

      // Write keyword arguments
      Object.entries(this.kwargs).forEach(([key, value], index, array) => {
        writer.write(`${key}=${value}`);
        if (index < array.length - 1) {
          writer.write(", ");
        }
      });

      writer.write(")");
    }
  }
}
