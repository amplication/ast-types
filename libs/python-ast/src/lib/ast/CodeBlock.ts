import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { Import } from "./Import";

/**
 * Configuration arguments for creating a Python code block.
 */
export type CodeBlockArgs =
  | {
      /** The code to write */
      code: string;
      /** A list of imports that are used in the code */
      imports?: Import[] | null;
    }
  | {
      /** A function that writes code to the writer */
      code: (writer: Writer) => void;
      imports?: never;
    };

/**
 * Represents a block of Python code in the AST.
 * This class handles the generation of arbitrary Python code blocks, either as
 * static strings or as dynamic code generation functions. It is used for language
 * features not directly supported by specific AST node classes.
 *
 * @extends {AstNode}
 */
export class CodeBlock extends AstNode {
  /** The code content or code generation function */
  private value: string | ((writer: Writer) => void);
  /** The imports used in the code */
  private imports: Import[];

  /**
   * Creates a new Python code block.
   * @param {CodeBlockArgs} args - The configuration arguments
   */
  constructor(args: CodeBlockArgs) {
    super();
    this.value = args.code;
    this.imports = [];

    if ("imports" in args && args.imports) {
      this.imports.push(...args.imports);
    }
  }

  /**
   * Writes the code block to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    if (typeof this.value === "string") {
      this.imports.forEach((import_) => writer.addImport(import_));
      writer.write(this.value);
    } else {
      this.value(writer);
    }
  }
}
