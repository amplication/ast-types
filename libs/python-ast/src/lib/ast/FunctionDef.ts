import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { ClassReference } from "./ClassReference";
import { CodeBlock } from "./CodeBlock";
import { Decorator } from "./Decorator";
import { Parameter } from "./Parameter";
import { Return } from "./Return";

/**
 * Configuration arguments for creating a Python function.
 */
export interface FunctionDefArgs {
  /** The name of the function */
  name: string;
  /** The parameters of the function */
  parameters?: Parameter[];
  /** Decorators to apply to the function */
  decorators?: Decorator[];
  /** Documentation string for the function */
  docstring?: string;
  /** Return type annotation */
  returnType?: ClassReference;
  /** Whether this is a static method */
  isStatic?: boolean;
  /** Whether this is a class method */
  isClassMethod?: boolean;
  /** Whether this is an async function */
  isAsync?: boolean;
}

/**
 * Represents a Python function definition in the AST.
 * This class handles the generation of Python function declarations including
 * parameters, decorators, and return type annotations.
 *
 * @extends {AstNode}
 */
export class FunctionDef extends AstNode {
  /** The name of the function */
  public readonly name: string;
  /** The parameters of the function */
  public readonly parameters: Parameter[];
  /** Decorators applied to the function */
  public readonly decorators: Decorator[];
  /** The documentation string */
  public readonly docstring?: string;
  /** The return type annotation */
  public readonly returnType?: ClassReference;
  /** Whether this is a static method */
  public readonly isStatic: boolean;
  /** Whether this is a class method */
  public readonly isClassMethod: boolean;
  /** Whether this is an async function */
  public readonly isAsync: boolean;

  /** The body of the function */
  private body: (CodeBlock | Return)[] = [];

  /**
   * Creates a new Python function definition.
   * @param {FunctionDefArgs} args - The configuration arguments
   */
  constructor({
    name,
    parameters = [],
    decorators = [],
    docstring,
    returnType,
    isStatic = false,
    isClassMethod = false,
    isAsync = false,
  }: FunctionDefArgs) {
    super();
    this.name = name;
    this.parameters = parameters;
    this.decorators = [...decorators];
    this.docstring = docstring;
    this.returnType = returnType;
    this.isStatic = isStatic;
    this.isClassMethod = isClassMethod;
    this.isAsync = isAsync;

    // Add appropriate decorators for static and class methods
    if (
      this.isStatic &&
      !this.decorators.some((d) => d.name === "staticmethod")
    ) {
      this.decorators.push(new Decorator({ name: "staticmethod" }));
    }
    if (
      this.isClassMethod &&
      !this.decorators.some((d) => d.name === "classmethod")
    ) {
      this.decorators.push(new Decorator({ name: "classmethod" }));
    }
  }

  /**
   * Adds a statement to the function body.
   * @param {CodeBlock | Return} statement - The statement to add
   */
  public addStatement(statement: CodeBlock | Return): void {
    this.body.push(statement);
  }

  /**
   * Writes the function definition and its body to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    // Write decorators
    this.decorators.forEach((decorator) => {
      decorator.write(writer);
      writer.newLine();
    });

    // Write function definition
    if (this.isAsync) {
      writer.write("async ");
    }
    writer.write(`def ${this.name}(`);

    // Write parameters
    this.parameters.forEach((param, index) => {
      param.write(writer);
      if (index < this.parameters.length - 1) {
        writer.write(", ");
      }
    });

    writer.write(")");

    // Write return type annotation if provided
    if (this.returnType) {
      writer.write(" -> ");
      this.returnType.write(writer);
    }

    writer.write(":");
    writer.newLine();
    writer.indent();

    // Write docstring if provided
    if (this.docstring) {
      writer.writeLine(`"""${this.docstring}"""`);
      if (this.body.length > 0) {
        writer.newLine();
      }
    }

    // Write function body
    if (this.body.length > 0) {
      this.body.forEach((statement) => {
        statement.write(writer);
        writer.newLine();
      });
    } else {
      // Empty function body needs a pass statement
      writer.writeLine("pass");
    }

    writer.dedent();
  }
}
