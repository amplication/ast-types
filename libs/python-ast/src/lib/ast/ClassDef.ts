import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { Assign } from "./Assign";
import { ClassReference } from "./ClassReference";
import { CodeBlock } from "./CodeBlock";
import { Decorator } from "./Decorator";
import { FunctionDef } from "./FunctionDef";

/**
 * Configuration arguments for creating a Python class.
 */
export interface ClassDefArgs {
  /** The name of the class */
  name: string;
  /** The module name containing the class */
  moduleName?: string;
  /** The parent classes to inherit from */
  bases?: ClassReference[];
  /** Decorators to apply to the class */
  decorators?: Decorator[];
  /** Documentation string for the class */
  docstring?: string;
}

/**
 * Represents a Python class definition in the AST.
 * This class handles the generation of Python class declarations including methods,
 * attributes, and nested classes.
 *
 * @extends {AstNode}
 */
export class ClassDef extends AstNode {
  /** The name of the class */
  public readonly name: string;
  /** The module name containing the class */
  public readonly moduleName?: string;
  /** The parent classes to inherit from */
  public readonly bases: ClassReference[];
  /** The decorators applied to the class */
  public readonly decorators: Decorator[];
  /** The documentation string for the class */
  public readonly docstring?: string;
  /** The class reference for this class */
  public readonly reference: ClassReference;

  /** The methods defined in this class */
  private methods: FunctionDef[] = [];
  /** The attributes defined in this class */
  private attributes: Assign[] = [];
  /** The nested classes defined in this class */
  private nestedClasses: ClassDef[] = [];
  /** Custom code blocks for unsupported features */
  private codeBlocks: CodeBlock[] = [];

  /**
   * Creates a new Python class definition.
   * @param {ClassDefArgs} args - The configuration arguments
   */
  constructor({
    name,
    moduleName,
    bases = [],
    decorators = [],
    docstring,
  }: ClassDefArgs) {
    super();
    this.name = name;
    this.moduleName = moduleName;
    this.bases = bases;
    this.decorators = decorators;
    this.docstring = docstring;

    this.reference = new ClassReference({
      name: this.name,
      moduleName: this.moduleName,
    });
  }

  /**
   * Adds a method to the class.
   * @param {FunctionDef} method - The method to add
   */
  public addMethod(method: FunctionDef): void {
    this.methods.push(method);
  }

  /**
   * Adds an attribute to the class.
   * @param {Assign} attribute - The attribute to add
   */
  public addAttribute(attribute: Assign): void {
    this.attributes.push(attribute);
  }

  /**
   * Adds a nested class to this class.
   * @param {ClassDef} nestedClass - The nested class to add
   */
  public addNestedClass(nestedClass: ClassDef): void {
    this.nestedClasses.push(nestedClass);
  }

  /**
   * Adds a code block for unsupported features.
   * @param {CodeBlock} codeBlock - The code block to add
   */
  public addCodeBlock(codeBlock: CodeBlock): void {
    this.codeBlocks.push(codeBlock);
  }

  /**
   * Writes the class definition and its members to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    // Write decorators
    this.decorators.forEach((decorator) => {
      decorator.write(writer);
      writer.newLine();
    });

    // Write class definition
    writer.write(`class ${this.name}`);

    // Write base classes if any
    if (this.bases.length > 0) {
      writer.write("(");
      this.bases.forEach((base, index) => {
        base.write(writer);
        if (index < this.bases.length - 1) {
          writer.write(", ");
        }
      });
      writer.write(")");
    }

    writer.write(":");
    writer.newLine();
    writer.indent();

    // Write docstring if provided
    if (this.docstring) {
      writer.writeLine(`"""${this.docstring}"""`);
      if (
        this.attributes.length > 0 ||
        this.methods.length > 0 ||
        this.nestedClasses.length > 0 ||
        this.codeBlocks.length > 0
      ) {
        writer.newLine();
      }
    }

    // Write attributes
    if (this.attributes.length > 0) {
      this.attributes.forEach((attribute) => {
        attribute.write(writer);
        writer.newLine();
      });
      if (
        this.methods.length > 0 ||
        this.nestedClasses.length > 0 ||
        this.codeBlocks.length > 0
      ) {
        writer.newLine();
      }
    }

    // Write methods
    this.methods.forEach((method, index, array) => {
      method.write(writer);
      if (index < array.length - 1) {
        writer.newLine();
      }
    });

    if (
      this.methods.length > 0 &&
      (this.nestedClasses.length > 0 || this.codeBlocks.length > 0)
    ) {
      writer.newLine();
      writer.newLine();
    }

    // Write nested classes
    this.nestedClasses.forEach((nestedClass, index, array) => {
      nestedClass.write(writer);
      if (index < array.length - 1) {
        writer.newLine();
        writer.newLine();
      }
    });

    if (this.nestedClasses.length > 0 && this.codeBlocks.length > 0) {
      writer.newLine();
      writer.newLine();
    }

    // Write additional code blocks
    this.codeBlocks.forEach((codeBlock, index, array) => {
      codeBlock.write(writer);
      if (index < array.length - 1) {
        writer.newLine();
      }
    });

    // If the class is empty, add a pass statement
    if (
      this.attributes.length === 0 &&
      this.methods.length === 0 &&
      this.nestedClasses.length === 0 &&
      this.codeBlocks.length === 0 &&
      !this.docstring
    ) {
      writer.writeLine("pass");
    }

    writer.dedent();
  }
}
