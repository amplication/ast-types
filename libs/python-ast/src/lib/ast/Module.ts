import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { ClassDef } from "./ClassDef";
import { CodeBlock } from "./CodeBlock";
import { FunctionDef } from "./FunctionDef";
import { Import } from "./Import";

/**
 * Configuration arguments for creating a Python module.
 */
export interface ModuleArgs {
  /** The name of the module */
  name: string;
  /** The docstring for the module */
  docstring?: string;
}

/**
 * Represents a Python module in the AST.
 * This class is the top-level container for Python code, including imports,
 * functions, classes, and global code.
 *
 * @extends {AstNode}
 */
export class Module extends AstNode {
  /** The name of the module */
  public readonly name: string;
  /** The docstring for the module */
  public readonly docstring?: string;

  /** The imports defined in this module */
  private imports: Import[] = [];
  /** The children nodes (functions, classes, code blocks) in order of addition */
  private children: (FunctionDef | ClassDef | CodeBlock)[] = [];

  /**
   * Creates a new Python module.
   * @param {ModuleArgs} args - The configuration arguments
   */
  constructor({ name, docstring }: ModuleArgs) {
    super();
    this.name = name;
    this.docstring = docstring;
  }

  /**
   * Adds an import statement to the module.
   * @param {Import} importStatement - The import statement to add
   */
  public addImport(importStatement: Import): void {
    this.imports.push(importStatement);
  }

  /**
   * Adds a function to the module.
   * @param {FunctionDef} functionDef - The function to add
   */
  public addFunction(functionDef: FunctionDef): void {
    this.children.push(functionDef);
  }

  /**
   * Adds a class to the module.
   * @param {ClassDef} classDef - The class to add
   */
  public addClass(classDef: ClassDef): void {
    this.children.push(classDef);
  }

  /**
   * Adds a global code block to the module.
   * @param {CodeBlock} codeBlock - The code block to add
   */
  public addCodeBlock(codeBlock: CodeBlock): void {
    this.children.push(codeBlock);
  }

  /**
   * Writes the module and its contents to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    // Write imports
    if (this.imports.length > 0) {
      this.imports.forEach((import_) => import_.write(writer));
      writer.newLine();
    }

    // Write module docstring if provided
    if (this.docstring) {
      writer.writeLine(`"""${this.docstring}"""`);
      writer.newLine();
    }

    // Write all children in order
    if (this.children.length > 0) {
      this.children.forEach((child, index) => {
        child.write(writer);
        if (index < this.children.length - 1) {
          writer.newLine();
          writer.newLine();
        }
      });
    }
  }
}
