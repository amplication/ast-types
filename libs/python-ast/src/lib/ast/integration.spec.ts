import { Module } from "./Module";
import { ClassDef } from "./ClassDef";
import { FunctionDef } from "./FunctionDef";
import { Parameter } from "./Parameter";
import { ClassReference } from "./ClassReference";
import { Decorator } from "./Decorator";
import { CodeBlock } from "./CodeBlock";
import { Return } from "./Return";
import { Assign } from "./Assign";
import { Import } from "./Import";

describe("Python AST Integration", () => {
  it("should generate a complete, realistic Python file", () => {
    // Create the module with docstring
    const module = new Module({
      name: "user_model",
      docstring:
        "Module containing the User data model and related functionality.",
    });

    // Add imports in a single block to avoid duplicates
    const imports = [
      new Import({ moduleName: "typing", names: ["List", "Optional", "Dict"] }),
      new Import({ moduleName: "datetime", names: ["datetime"] }),
      new Import({ moduleName: "uuid", names: ["UUID"] }),
      new Import({ moduleName: "dataclasses", names: ["dataclass", "field"] }),
    ];

    imports.forEach((imp) => module.addImport(imp));

    // Create a data model class
    const userClass = new ClassDef({
      name: "User",
      decorators: [
        new Decorator({ name: "dataclass", moduleName: "dataclasses" }),
      ],
      docstring:
        "Represents a user in the system with their associated metadata.",
    });

    // Add class variables with type annotations
    userClass.addAttribute(
      new Assign({
        target: "id",
        type: new ClassReference({ name: "UUID", moduleName: "uuid" }),
        value: "UUID.uuid4()",
        isClassVariable: true,
      }),
    );

    userClass.addAttribute(
      new Assign({
        target: "username",
        type: new ClassReference({ name: "str" }),
        value: "''",
        isClassVariable: true,
      }),
    );

    userClass.addAttribute(
      new Assign({
        target: "email",
        type: new ClassReference({ name: "str" }),
        value: "''",
        isClassVariable: true,
      }),
    );

    userClass.addAttribute(
      new Assign({
        target: "created_at",
        type: new ClassReference({ name: "datetime", moduleName: "datetime" }),
        value: "datetime.now()",
        isClassVariable: true,
      }),
    );

    userClass.addAttribute(
      new Assign({
        target: "roles",
        type: new ClassReference({ name: "List", moduleName: "typing" }),
        value: "field(default_factory=list)",
        isClassVariable: true,
      }),
    );

    userClass.addAttribute(
      new Assign({
        target: "metadata",
        type: new ClassReference({
          name: "Dict",
          moduleName: "typing",
        }),
        value: "field(default_factory=dict)",
        isClassVariable: true,
      }),
    );

    // Add a class method for creating a user
    const createMethod = new FunctionDef({
      name: "create",
      isClassMethod: true,
      parameters: [
        new Parameter({ name: "cls" }),
        new Parameter({
          name: "username",
          type: new ClassReference({ name: "str" }),
        }),
        new Parameter({
          name: "email",
          type: new ClassReference({ name: "str" }),
        }),
        new Parameter({
          name: "roles",
          type: new ClassReference({
            name: "List",
            moduleName: "typing",
          }),
          default_: "None",
        }),
      ],
      returnType: new ClassReference({ name: "User" }),
    });

    createMethod.addStatement(
      new CodeBlock({
        code: `return cls(
    id=UUID.uuid4(),
    username=username,
    email=email,
    roles=roles or [],
    created_at=datetime.now(),
    metadata={})`,
      }),
    );

    userClass.addMethod(createMethod);

    // Add an instance method
    const addRoleMethod = new FunctionDef({
      name: "add_role",
      parameters: [
        new Parameter({ name: "self" }),
        new Parameter({
          name: "role",
          type: new ClassReference({ name: "str" }),
        }),
      ],
      returnType: new ClassReference({ name: "None" }),
    });

    addRoleMethod.addStatement(
      new CodeBlock({
        code: "if role not in self.roles:\n    self.roles.append(role)",
      }),
    );

    userClass.addMethod(addRoleMethod);

    // Add a property
    const isAdminProperty = new FunctionDef({
      name: "is_admin",
      decorators: [new Decorator({ name: "property" })],
      parameters: [new Parameter({ name: "self" })],
      returnType: new ClassReference({ name: "bool" }),
    });

    isAdminProperty.addStatement(
      new Return({ value: "'admin' in self.roles" }),
    );

    userClass.addMethod(isAdminProperty);

    // Add the class to the module
    module.addClass(userClass);

    // Add example usage as a comment at the end of the file
    module.addCodeBlock(
      new CodeBlock({
        code: `

# Example usage:
if __name__ == "__main__":
    user = User.create("john_doe", "john@example.com")
    user.add_role("admin")
    print(f"Is admin: {user.is_admin}")
`,
      }),
    );

    // Generate the complete file
    const output = module.toString();

    // Compare with snapshot
    expect(output).toMatchSnapshot();

    // Additional specific assertions to ensure correct structure
    const lines = output.split("\n");

    // Check imports are at the top and not duplicated
    const importLines = lines.filter(
      (line) => line.startsWith("from") || line.startsWith("import"),
    );
    const uniqueImports = new Set(importLines);
    expect(importLines.length).toBe(uniqueImports.size);

    // Check docstring is after imports
    const docstringIndex = lines.findIndex((line) =>
      line.includes('"""Module containing'),
    );
    const lastImportIndex = lines.reduce(
      (max, line, index) =>
        line.startsWith("from") || line.startsWith("import") ? index : max,
      -1,
    );
    expect(docstringIndex).toBeGreaterThan(lastImportIndex);

    // Check class is after docstring
    const classIndex = lines.findIndex((line) => line.includes("class User:"));
    expect(classIndex).toBeGreaterThan(docstringIndex);

    // Check example is at the end
    const exampleIndex = lines.findIndex((line) =>
      line.includes("# Example usage:"),
    );
    expect(exampleIndex).toBeGreaterThan(classIndex);

    // Check for specific content
    expect(output).toContain("from typing import Dict, List, Optional");
    expect(output).toContain("from datetime import datetime");
    expect(output).toContain("from uuid import UUID");
    expect(output).toContain("from dataclasses import dataclass, field");
    expect(output).toContain("@dataclass");
    expect(output).toContain("class User:");
    expect(output).toContain("def create(");
    expect(output).toContain("def add_role(");
    expect(output).toContain("@property");
    expect(output).toContain("def is_admin(");
  });
});
