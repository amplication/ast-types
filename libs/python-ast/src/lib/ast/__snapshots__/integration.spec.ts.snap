// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Python AST Integration should generate a complete, realistic Python file 1`] = `
"from dataclasses import dataclass, field
from datetime import datetime
from typing import Dict, List, Optional
from uuid import UUID


"""Module containing the User data model and related functionality."""

@dataclass
class User:
    """Represents a user in the system with their associated metadata."""

    id: UUID = UUID.uuid4()
    username: str = ''
    email: str = ''
    created_at: datetime = datetime.now()
    roles: List = field(default_factory=list)
    metadata: Dict = field(default_factory=dict)

    @classmethod
    def create(cls, username: str, email: str, roles: List = None) -> User:
        return cls(
            id=UUID.uuid4(),
            username=username,
            email=email,
            roles=roles or [],
            created_at=datetime.now(),
            metadata={})

    def add_role(self, role: str) -> None:
        if role not in self.roles:
            self.roles.append(role)

    @property
    def is_admin(self) -> bool:
        return 'admin' in self.roles




# Example usage:
if __name__ == "__main__":
    user = User.create("john_doe", "john@example.com")
    user.add_role("admin")
    print(f"Is admin: {user.is_admin}")
"
`;
